vochain/state/sik.go

```diff
@@ -0,0 +1,247 @@
+// height provided, ready to use in the SIK subTree as leaf value to invalidate
+// it. The encoded value will have 32 bytes:
+//   - The initial 28 bytes must be zero.
+//   - The remaining 4 bytes must contain the height encoded in LittleEndian
+func (s SIK) InvalidateAt(height uint32) SIK {
+	bHeight := big.NewInt(int64(height)).Bytes()
+	// fill with zeros until reach the encoded height length
+	for len(bHeight) < encodedHeightLen {
+		bHeight = append([]byte{0}, bHeight...)
+	}
+	// create the encodedHeight with the right number of zeros
+	s = make([]byte, sikLeafValueLen-encodedHeightLen)
+	// copy the height bytes swapping endianness in the last bytes
+	for i := encodedHeightLen - 1; i >= 0; i-- {
+		s = append(s, bHeight[i])
+	}
+	return s
+}
+
+// DecodeInvalidatedHeight funtion returns the decoded height uint32 from the
+// leaf value that contains an invalidated SIK.
+func (s SIK) DecodeInvalidatedHeight() uint32 {
+	bHeight := []byte{}
+	for i := sikLeafValueLen - 1; len(bHeight) < encodedHeightLen; i-- {
+		bHeight = append(bHeight, s[i])
+	}
+	return uint32(new(big.Int).SetBytes(bHeight).Int64())
+}
+
+// Valid function returns if the current SIK is a valid one or not.
+func (s SIK) Valid() bool {
+	for i := 0; i < len(s)-encodedHeightLen; i++ {
+		if s[i] != 0 {
+			return true
+		}
+	}
+	return false
+}
+
+// String function return the human readable version of the current SIK, if it
+// is a valid one, return the SIK value as hex. If it is already invalidated,
+// return the decoded height.
+func (s SIK) String() string {
+	if s.Valid() {
+		return hex.EncodeToString(s)
+	}
+	return fmt.Sprint(s.DecodeInvalidatedHeight())
+}
+
+// ValidSIKRoots return the list of current valid roots from the SIK's merkle
+// tree. It reads the roots from the key-value database associated to the SIK's
+// subtree.
+func (v *State) ValidSIKRoots() ([][]byte, error) {
+	v.Tx.Lock()
+	defer v.Tx.Unlock()
+	siksTree, err := v.Tx.DeepSubTree(StateTreeCfg(TreeProcess))
+	if err != nil {
+		return nil, err
+	}
+	validRoots := [][]byte{}
+	siksTree.NoState().Iterate(nil, func(_, root []byte) bool {
+		validRoots = append(validRoots, root)
+		return true
+	})
+	return validRoots, nil
+}
+
+// UpdateSIKRoots keep on track the last valid SIK Merkle Tree roots to support
+// voting to already registered users when an election is on going and new users
+// are registered. When a new sikRoot is generated, the sikRoot’s from an older
+// block than the current block minus the hysteresis blocks will be deleted:
+//   - If exists a sikRoot for the minimun hysteresis block number
+//     (currentBlock - hysteresis), just remove all the roots with a lower block
+//     number.
+//   - If it does not exist, remove all roots with a lower block number except
+//     for the next lower sikRoot. It is becouse it still being validate for a
+//     period.
+func (v *State) UpdateSIKRoots() error {
```

Comment:
I wonder if for SIKs we can follow the same approach as in ProcessBlockRegistry. 
My feeling is that the State package is growing too much and some hierarchy would be appreciated.